Source maps applied to wasm binaries

dschuff opened the pull request:
This PR proposes applying source maps to wasm binaries. Source maps obviously have a lot of known limitations and are not a long-term debugging solution. However they are widely supported (including by emscripten for asm.js code) and based on conversations with Chrome and Firefox dev tools folks, my hope is that it will be easy and straightforward to adapt the existing dev tools code to apply to wasm binaries in the meantime, and bring wasm to parity with asm.js.

kanaka commented:
Couldn't the wasm be disassembled to wast textual format and the links to the original source be attached to lines in the wast (rather than an offset into the wasm binary)? Or perhaps both?

binji commented on Web.md:
```diff
@@ -154,6 +154,78 @@ Transcoding failure is detected by `decodeURIComponent`, which may throw
+generate a stack trace). However existing developer tools implementations are
+designed around the source map model of passing a URL from the VM to the dev
+tools code, which fetches and interprets the information. Therefore it is
+epxected that it will be much simpler to use "real" source maps than to design
```
sp: expected

binji commented on Web.md:
```diff
@@ -154,6 +154,78 @@ Transcoding failure is detected by `decodeURIComponent`, which may throw
+
+Source maps specify how to map a (zero-based) line and column position in
+generated code to a file, line, and column location in the source. For
+wasm binary locations, the line number is always 0, and the column number
```
You could use the generated "line number" as the function index instead (or perhaps just the defined function index, skipping imported functions), and the generated "column" as the instruction offset in the function. But I suppose it's best to align w/ what's being discussed [here](https://github.com/WebAssembly/design/issues/990).

Is it worth mentioning this in some rationale?

dschuff replied:
Hm, On one hand the source map itself is not really meant to be read by humans. But if some tool displays it, then yeah I think it would make sense to display the offsets the way they'd be displayed in tools or browsers, yeah. And yeah it probably is worth mentioning. Maybe I'll do a PR for #990 too and add that to the rationale.

yurydelendik replied:
Using column number (vs line number) in JS source maps gives a really good savings, to be precise the size of the wasm bytes (the `;` used to separate every encoded line and since we will have only one line, smaller number of column separators will be used)

yurydelendik replied:
We cannot use line number == 0 due to map encoding specifics. And 'source-map' library throws when line == 1.

dschuff replied:
Fixed. I had thought I used 0 in my original prototype, but it was actually 1.

dschuff commented:
@kanaka IIRC In Chrome and Firefox, currently when displaying wast (instead of original source) the JS engine disassembles the wasm binary and passes the wast text to the devtools. Using source maps would mean the source would be displayed instead of the wast. So in that case you wouldn't need the wast at all.
Are you suggesting that the wast be generated offline, and then passed to the browser along with a source map? In that case we'd still need a way to link from the wasm binary to the wast. In that sense the wast would be sort of like a mapped source for the binary. But it wouldn't really save much because browsers would probably still want to have some support for the case where there is nothing but the binary. Which means they'd probably still have to support disassembling it themselves, so I would guess that passing the disassembly from the outside wouldn't really save anything or make things any simpler.

RyanLamansky commented:
I was thinking a binary source map format designed for WASM specifically would be more desirable than re-using the JavaScript format.  But I'm not a browser vendor so I'm not sure how troublesome that might be.

dschuff commented:
@RyanLamansky I agree that in principle a source-mapping format meant for wasm would not look like JavaScript source maps. However the intention of this convention is just to capture some really low-hanging fruit, because what we ultimately want is something much more capable than sourcemaps, and that will either have to be designed independently, or re-used from some other existing format (e.g. DWARF). So in one sense, this isn't worth doing at all if it isn't really simple, because it will be superseded, hopefully before too long. Therefore it should fit best with what browsers or devtools are already doing with source maps, which means that it would be nice not to require them to parse a format which is new, but also not the final form that we want.

yurydelendik commented on Web.md:
```diff
@@ -154,6 +154,78 @@ Transcoding failure is detected by `decodeURIComponent`, which may throw
+the URL is defined as in [RFC3986](https://tools.ietf.org/html/rfc3986) (e.g.
+it must be percent-encoded if necessary) and it may be a data URI. The URL
+is also resolved according to the source map spec, and may also be specified
+with the `SourceMap:` HTTP header.
```
Is there a spec/reference to how HTTP headers will be associated with wasm module (assuming it's coming from Response)?

lukewagner replied:
I would think this would only work for the new `Response`-taking `compile` and `instantiate` overloads.

jfbastien commented on Web.md:
```diff
@@ -218,11 +218,13 @@ Source locations are interpreted as in the source map spec.
-it must be percent-encoded if necessary) and it may be a data URI. The URL
-is also resolved according to the source map spec, and may also be specified
-with the `SourceMap:` HTTP header.
+The URL is defined as in the the WHATWG
```
the the

jfbastien commented:
Could you also update https://github.com/WebAssembly/design/blob/master/FutureFeatures.md#source-maps-integration ?

flagxor commented:
Pinging this back to life :-)
Looks like this is waiting on some fixes from Derek.

dschuff commented:
OK, I think have addressed the comments, although I'm not 100% sure if @domenic had anything more in mind beyond what I did about URLs and resolution.

domenic commented:
It seems reasonable given the general level of detail here. Thanks!

yurydelendik commented:
Demo compiled using emscripten that is using source maps https://yurydelendik.github.io/sqlite-playground/src/playground.html (source at https://github.com/yurydelendik/sqlite-playground)

yurydelendik commented on Web.md:
```diff
@@ -299,6 +299,86 @@ Transcoding failure is detected by `decodeURIComponent`, which may throw
+compileStreaming or instantiateStreaming
+[APIs](#additional-web-embedding-api) are used), then it is resolved relative to
+the URL of the response.
+* Otherwise it is relative to the page's origin.
```
Just to understand the relative resolution with "page's origin", if `WebAssembly.instantiate()` was called from a page's script at "http://example.org/~user/page.html", and wasm's `sourceMappingURL` is specified as "maps/t.wasm.map", the resulting source map URL will be "http://example.org/maps/t.wasm.map", is it correct? Does `document.baseURI` affect this?

littledan commented:
I'm wondering if we should land this patch in the source map spec rather than in the WebAssembly spec. I filed [an issue](https://github.com/source-map/source-map-rfc/issues/8) in the source map spec's tracker to follow this. JavaScript and CSS leave source maps to be described in the source map spec, so why should WebAssembly be different here?

dschuff commented:
I started the discussion here to get feedback from wasm folks but I agree that it makes more sense to go in the source map spec, and wasm shouldn't be different from JS.

fitzgen commented:
> source map spec's tracker

FYI, that RFC repo never caught on, and I ultimately could never get anyone (authors of compilers targeting JS, browser devtools implementors) to participate at the time.

Since source maps are a stop gap for now, but ultimately a dead end, I'd suggest two things:

1. Just implementing and documenting the column=byte offset thing.

2. Starting a new effort to create a real debugging format (or adopting DWARF). I would be happy to participate in such an effort.

littledan commented:
@fitzgen Since sourcemaps are web reality, and it'd likely hurt debugging in the real world if we just removed them in favor of a new version, maybe we should both strengthen the specification and conformance tests for existing sourcemaps as well as adding a v4 version. We'll need to get all the stakeholders involved whether we just do a new version or whether we also improve the specification for the previous version, anyway.

Would people be interested in getting together for a call to discuss rebooting the sourcemaps spec?

cc @foolip @ak239 @concavelenz

foolip commented:
I have no previous involvement with sourcemaps, maybe @mathiasbynens knows who might know?

dschuff commented:
I'm definitely interested in moving debugging forward. This particular proposal is just a small adaptation to the existing scope and functionality of source maps (i.e. as purely a line-table facility but not full debug info), so I'm happy to put it wherever people thing it belongs, and contribute tests and docs where appropriate.

For "real" debugging, that's an entirely new class of capabilities (even in DWARF, line tables are a separate section and fairly independent from the rest of the debug info). If there's a new data format (regardless of whether it's sourcemap-like or DWARF-like or something else) it could be pretty independent too.

Beyond that though, I'm actually not fully convinced that a file format is even the best thing to standardize on as the primary interoperability layer; something like an API or protocol might have some advantages; but I guess exactly what that means can be discussed somewhere other than this PR.

sjrd commented:
As the author of Scala.js, I would be interested to be part of discussions about improving source maps. In our community we are painfully aware of a number of shortcomings of source maps.

ak239 commented:
On DevTools side we think that current static format of source maps can be not suitable for WebAssembly and we are looking for some dynamic format. I can imagine some DWARF based ideas, as well as something similar with language services or adapter over well supported by industry DevTools protocol.

littledan commented:
Would folks here be interested in attending a video call in the next month or two to get started on improving debugging metadata, for both Wasm and JS?

xtuc commented:
I would be interested @littledan, even if I didn't participate much here?

gskachkov commented:
@littledan In my own internal planes for this year, is participating in adding source map and debugging to wasm in JSC, so I would be interested. Are there any free places left?

titzer commented:
I agree with @dschuff. Instead of a file format, we should standard debugging protocol API (similar to JVMTI or JDWP) which would give programmatic access to the WASM-machine-level state and give control for stopping, stepping, etc. Then DWARF and any other file format can be processed in a "user space" tool that just interacts with the WebAssembly engine via this standardized protocol.

IMO this has the advantage of keeping the WebAssembly engine completely independent from any particular debugging format, making it maximally future-proof. (I also think it would be the simplest thing to implement at the engine layer).

tschneidereit commented:
I tend to agree that specifying this kind of debugging protocol API seems to make sense, but I know that @fitzgen had some good arguments against it that'd need to be discussed. I'm quite sure I'd butcher his arguments, so I won't try to represent them here.

Meanwhile, @yurydelendik is working on extensions to DWARF that would make it work for Wasm, with the intent to embed it into custom sections and synthesizing source maps from them on the fly. This is obviously not something we'll want to standardize in this form - or probably even ship in a release build. It is a good step towards validating the viability of using DWARF as the format for representing this information.

chicoxyzzy commented:
I would like to join video call if it's possible

fitzgen commented:
Hello all!

I am a little concerned that when the general subject of "debug info for wasm" comes up, folks are only thinking of the stepping debugger use case.

Here is a concrete list of debug info I wished that wasm had when I was writing tooling for wasm or using wasm tooling:

* Which logical functions are inlined within a given physical function? Within the function's body, which byte ranges can be attributed to each inlined function, and which to the physical function itself?

  * My first use case for this information is in the [`twiggy` code size profiler][twiggy], where I want to find large functions that are getting inlined many times, and might be causing bloat.

  * My second use case is: when profiling WebAssembly code, sampled stack frames are *much* less useful than their JavaScript counterparts, because only physical frames are displayed. Profilers should display inlined function's frames as well, like they do for JavaScript (with assistance from the JS engine, rather than debug info).

  * As a precedent, this information is present in DWARF's `.debug_info` section, as a `DW_TAG_inlined_subroutine` entry in the debugging information entry (DIE) tree:

    <details>

    ```
    < 4><0x00000454>          DW_TAG_subprogram
                                DW_AT_low_pc                0x00016d30
                                DW_AT_high_pc               <offset-from-lowpc>635
                                DW_AT_frame_base            len 0x0001: 56: DW_OP_reg6
                                DW_AT_linkage_name          _ZN12cpp_demangle3ast11BuiltinType8demangle17h174c82f68696be09E
                                DW_AT_name                  demangle<&mut alloc::vec::Vec<u8>>
                                DW_AT_decl_file             0x00000001 /home/fitzgen/cpp_demangle/src/ast.rs
                                DW_AT_decl_line             0x00000c35
                                DW_AT_type                  <0x00016a94>
    ...
    < 6><0x000004ca>              DW_TAG_lexical_block
                                    DW_AT_low_pc                0x00016d51
                                    DW_AT_high_pc               <offset-from-lowpc>596
    < 7><0x000004d7>                DW_TAG_variable
                                      DW_AT_location              <loclist at offset 0x00009e83 with 1 entries follows>
                [ 0]< offset pair low-off : 0x00016d4b addr  0x00016d4b high-off  0x00016d59 addr 0x00016d59>DW_OP_breg4+1 DW_OP_stack_value
                                      DW_AT_name                  ty
                                      DW_AT_alignment             0x00000001
                                      DW_AT_decl_file             0x00000001 /home/fitzgen/cpp_demangle/src/ast.rs
                                      DW_AT_decl_line             0x00000c40
                                      DW_AT_type                  <0x00018731>
    < 7><0x000004e8>                DW_TAG_inlined_subroutine
                                      DW_AT_abstract_origin       <0x00000636>
                                      DW_AT_low_pc                0x00016d51
                                      DW_AT_high_pc               <offset-from-lowpc>596
                                      DW_AT_call_file             0x00000001 /home/fitzgen/cpp_demangle/src/ast.rs
                                      DW_AT_call_line             0x00000c40
    < 8><0x000004fc>                  DW_TAG_inlined_subroutine
                                        DW_AT_abstract_origin       <0x00016893>
                                        DW_AT_low_pc                0x00016f63
                                        DW_AT_high_pc               <offset-from-lowpc>58
                                        DW_AT_call_file             0x00000003 /home/fitzgen/cpp_demangle/<write macros>
                                        DW_AT_call_line             0x00000002
    < 9><0x0000050f>                    DW_TAG_formal_parameter
                                          DW_AT_location              len 0x0007: 930810019f9308: DW_OP_piece 8 DW_OP_constu 1 DW_OP_stack_value DW_OP_piece 8
                                          DW_AT_abstract_origin       <0x000168a4>
    < 9><0x0000051c>                    DW_TAG_formal_parameter
                                          DW_AT_location              <loclist at offset 0x00009ea8 with 2 entries follows>
                [ 0]< offset pair low-off : 0x00016f63 addr  0x00016f63 high-off  0x00016f8d addr 0x00016f8d>DW_OP_piece 8 DW_OP_constu 1 DW_OP_stack_value DW_OP_piece 8
                [ 1]< offset pair low-off : 0x00016f8d addr  0x00016f8d high-off  0x00016fab addr 0x00016fab>DW_OP_breg6-16 DW_OP_piece 8 DW_OP_constu 1 DW_OP_stack_value DW_OP_piece 8
                                          DW_AT_abstract_origin       <0x000168af>
    < 9><0x00000525>                    DW_TAG_formal_parameter
                                          DW_AT_location              len 0x0007: 930810019f9308: DW_OP_piece 8 DW_OP_constu 1 DW_OP_stack_value DW_OP_piece 8
                                          DW_AT_abstract_origin       <0x000168ba>
    ...
    ```

    </details>

* Is a given function a monomorphization of some generic function?

  * Again, I want this information in `twiggy`, so I can try and identify generic functions that have been monomorphized "too many" times and are leading to code bloat.

  * DWARF doesn't represent the generic functions themselves, but does have `DW_TAG_template_{type,value}_parameter` entries, so you can mostly reconstruct this information.

* Mapping source locations to wasm byte codes and back.

  * The one thing source maps can do :)

  * DWARF's equivalent is in the `.debug_line` section.

  * For `twiggy`, I want to say where in the source call graph edges are originating from. If you expect `blah_function` to be dead code, and removed by the linker, but it *isn't*, I want to be able to point at a series of "src/foo.rs:12:34 in `lolo_method`" as the reasons why the `blah_function` is not dead code and is included in the `.wasm` binary.

I want to highlight that this is just the concrete use cases I have run into *so far*. As we do more with wasm, this list will only grow. Any kind of instrumentation or static analysis tooling is going to want to consume debug info.

Why do I want to highlight debug info consumer use cases besides stepping debuggers? Because I want to make sure that you don't have to be running the debuggee program just to get information about it. A size profiler, a static analyzer, and an instrumentation-based code coverage tool shouldn't have to run the `.wasm` binary they are working on.

[twiggy]: https://github.com/rustwasm/twiggy

fitzgen commented:
@littledan I would love to join a video call to discuss this topic :)

codehag commented:
Let's set up a video call. Are people available next week? I will set up a doodle if I get enough thumbs up.
@fitzgen @littledan @xtuc @chicoxyzzy @gskachkov pinging you specifically since you expressed interest. everyone else is also invited!

sjrd commented:
I am also very interested, but already overbooked until May 8 included.

gskachkov commented:
Hi there, there is small question about _wasm debug_:
As **wasm** can be run on mobile devices and we can't run DevTool there, so **wasm** should support some kind **remote debug protocol**, that allows debug _wasm modules_ on mobile devices. 
Will **remote debug protocol** be part of the **wasm debug specification** or it will be handled by each browser by own remote debugging protocol?

tschneidereit commented:
@codehag please invite @yurydelendik to this call, too.

fitzgen commented:
> As wasm can be run on mobile devices and we can't run DevTool there, so wasm should support some kind remote debug protocol, that allows debug wasm modules on mobile devices.
> Will remote debug protocol be part of the wasm debug specification or it will be handled by each browser by own remote debugging protocol?

I expect wasm debug info and remote debug prototcols to be largely orthogonal concerns.

For example, the Firefox debugger used to apply source maps on the server side (ie the debuggee device) and has switched to applying them on the client side (ie the desktop displaying the UI). This tells me that they needn't be intertwined.

Did you have something in mind that requires them to be developed together?

ak239 commented:
On Chrome DevTools side, we think that JavaScript engine should provide low
level inspection and debugging tools using protocol, something like
Debugger and Runtime protocol domain for JavaScript but specific for
WebAssembly. Using this protocol tools would be able to do super basic
stuff, if tool needs something more fancy then it should fetch debugging
information from somewhere and use it to make everything nicer.

We think that this debugging information should be available for IDE not as
static format which should be parsed and consumed but as dynamic service
which will expose required API - we are working on proposal how this API
may look like right now. With this additional layer WebAssembly may
represent debugging information in any suitable for original language
format, maybe DWARF for C/C++ and SourceMap for something close to
JavaScript. Dynamic format will allow to resolve some well-known
limitations of existing sourcemap format, such as evaluating expressions in
original languages.

So we have basic protocol (works remotely as well) to do basic WebAssembly
specific inspection and debugging and we have separate service which IDE
may connect to and provide rich capabilities. This service internally uses
some sourcemap format which is suitable for original language and may be
developed by original language community.

I would be happy to join!

Thanks,
Aleksey.

On Wed, Apr 25, 2018 at 8:51 AM Nick Fitzgerald <notifications@github.com>
wrote:

> As wasm can be run on mobile devices and we can't run DevTool there, so
> wasm should support some kind remote debug protocol, that allows debug wasm
> modules on mobile devices.
> Will remote debug protocol be part of the wasm debug specification or it
> will be handled by each browser by own remote debugging protocol?
>
> I expect wasm debug info and remote debug prototcols to be largely
> orthogonal concerns.
>
> For example, the Firefox debugger used to apply source maps on the server
> side (ie the debuggee device) and has switched to applying them on the
> client side (ie the desktop displaying the UI). This tells me that they
> needn't be intertwined.
>
> Did you have something in mind that requires them to be developed together?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/WebAssembly/design/pull/1051#issuecomment-384336906>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAaBsoyJbyKOWbYsemdqDTH0tjzm2ubYks5tsJtwgaJpZM4NNmuA>
> .
>

gskachkov commented:
@fitzgen Thanks for answer, I see now that my question is not related to current thread.

@ak239  `offtopic` I'm just curious, if proposal for __dynamic service API__, that you spoke about, will be part of the TC39 or W3C spec?

codehag commented:
Hi everyone. I will close the doodle on wednesday, so please make sure to vote for a date.

In the meantime -- I have created an agenda, please comment if you want to add or remove something from the discussion: https://docs.google.com/document/d/1mMFN-OksiJZm1PD3XmBpAVq0vhT0df1rGKwYq2jfn58/edit#

codehag commented:
Hi everyone, based on the votes we have a tie between Monday and Wednesday. I would propose that we do Wednesday so that we have a little more time to prepare, @chicoxyzzy would you be able to make it?

chicoxyzzy commented:
@codehag I'm not sure, but I can try.

codehag commented:
@chicoxyzzy In case you cant make it, and if there is anything you want to ask / propose please add it to the doc and we will make sure to cover it, there will be notes

hopefully you can make it!

kumpera commented:
I'd love to join the call and Wednesday works great for me.

I'm currently working on doing C# debugging on top of WebAssembly and our current design aligns with the API proposals.

dschuff commented:
Hi Yulia, Sorry I've been out of office for a couple of weeks. I'm back now and any of those dates are fine for me.

fitzgen commented:
Hello everyone :)

In preparation for our meeting next week, I've compiled a list of requirements for wasm debugging capabilities. I've tried to focus solely on the kinds of tooling and queries we would like to support, and completely divorce these requirements from their eventual implementation (static format vs language server vs combo of those vs ...)  Rather than proposing a particular solution, this is an attempt to enumerate the requirements and constraints that *any* solution must satisfy.

This is what I've come up with: https://github.com/fitzgen/requirements-for-wasm-debugging-capabilities/blob/master/README.md

Let me know what I missed ;)

Looking forward to talking to y'all next week!

codehag commented:
Hey everyone, just a reminder -- meeting is at 7 pm CET (or 10:00pm Pacific Time (PT)) on May 9th, there was a wrong date in the doc that i linked.

codehag commented:
Sorry again, its 10 AM* pt

fitzgen commented:
Thanks again to everyone who came to the meeting today :)

I have completed my action item, a comparison between the requirements document and the ["First version of DebuggingFramework.md" pull request](https://github.com/WebAssembly/design/pull/708/). The comparison is here: https://github.com/fitzgen/requirements-for-wasm-debugging-capabilities/issues/2

ak239 commented:
I just summarized my ideas about API instead of static format: [here](https://github.com/ak239/webassembly-debug-information/blob/master/README.md).
Mine doc is much less detailed than Nick's and I mostly would like to share initial thoughts that we have inside Chrome DevTools team about WASM debugging.

auchenberg commented:
I'm late to the game in this thread. Kenneth here from VS Code. Our JavaScript debuggers are heavily dependent on sourcemaps, and we'd love to be a part of this conversation, and future meetings.

codehag commented:
@auchenberg hey kenneth! great! we will set up a dedicated communication channel for this. if you want to catch up with the meeting notes from wednesday, you can find them here: https://docs.google.com/document/d/1mMFN-OksiJZm1PD3XmBpAVq0vhT0df1rGKwYq2jfn58/edit#

fitzgen commented:
Thanks for sharing @ak239!

Before we go too deep exploring different designs, I think we should focus on defining requirements before proposing potential architectures and solutions. This way we all have a shared, common understanding of the problems we intend to solve, and we can weigh *any* proposed solution or architecture against these requirements.

Everyone: please read through the requirements and file issues or send PRs for things that you believe are missing! :)

> there are two type of clients: IDEs and static analysis tool

It seems that you are grouping browsers' developer tools (stepping debuggers, consoles, time profilers, etc) fall under the IDE umbrella, correct?

[I've enumerated a non-exhaustive, but hopefully representative set of tools we would like to support.](https://fitzgen.github.io/wasm-debugging-capabilities/#supported-tools) I don't currently have IDEs listed there, mostly because I'd like to get a better sense of which specific capabilities they need and what their requirements are, but I filed an [issue for that here](https://github.com/fitzgen/wasm-debugging-capabilities/issues/5). If there are any folks developing IDEs on this thread, it'd be *great* to get your comments in that issue!

--------------------------------------------------------------------------------

Here are some fairly unstructured thoughts while reading this, and thinking about the relationship between debugging capabilities and protocols:

* *Static* debugging capabilitities are orthogonal to developer tools wire protocols. Supporting evidence is that Firefox DevTools used to apply source maps in the protocol server, and switched to applying them on the protocol client.

  ```
  +---------------------+                           +-----------------------+
  | Server @ wasm-level | --- apply debug info ---> | Server @ source-level |
  +---------------------+                           +-----------------------+
            |                                                  |
            |                                                  |
    send over protocol                                 send over protocol
            |                                                  |
            V                                                  V
  +---------------------+                           +-----------------------+
  | Client @ wasm-level | --- apply debug info ---> | Client @ source-level |
  +---------------------+                           +-----------------------+
  ```

  Whether we apply debug info and then go over a protocol, or go over a protocol and then apply debug info we end up in the same place. Therefore, why constrain ourselves by tying debug info and DevTools protocols together?

* *Dynamic* debugging capabilities, on the other hand, *must* have direct access to the WebAssembly runtime. An extreme example: the only way to get the full list of frames currently on the stack of an interpreter written in Wasm (say Mono/C#) is to ask the running interpreter. A less extreme example: what is the current value of this arbitrary binding that is in scope?

* The DevTools protocol is Web-centric. I fear that it embeds many Web-isms and JS-isms, and that the debugging capabilities would not be useful for non-Web Wasm embedders. [2.1.2. Must be embedder agnostic](https://fitzgen.github.io/wasm-debugging-capabilities/#embedder-agnostic).

* It feels unnecessary for a static analysis tool or code size profiler to even speak *any* protocol. It has the `.wasm` binary right there, why shouldn't it also have the debugging capabilities right there?

* It seems that the primary benefit of using a protocol (or more specifically the Chrome DevTools protocol) is that there are a lot of existing consumers that we can give Wasm support with relatively little effort:

  > In modern ecosystem a lot of IDEs (e.g. VSCode, WebStorm) use DevTools protocol (mostly Runtime and Debugger domains) to debug JavaScript, so adding support for new WebAssembly domain should be easy for them.

  **However, we can realize this benefit without tying debugging capabilities to any protocol.** If the debugging capabilities are applied on the protocol server—as is necessary for dynamic debugging capabilities anyways—then all packets on the protocol will be speaking in terms of the original source, not the generated WebAssembly code. Therefore, all existing protocol clients receive the benefits of the debugging capabilities without any effort on their part.

* I believe this last point is similar to "Architecture B", but applying the debugging capabilities within the server, rather than as a protocol shim:

  ```
  +--------+                    +-------------------------------------+
  | Client | <--- protocol ---> | Server                              |
  +--------+                    |                                     |
                                |  +--------------+  +-------------+  |
                                |  | Debugging    |  | WebAssembly |  |
                                |  | Capabilities |  | Engine      |  |
                                |  +--------------+  +-------------+  |
                                |                                     |
                                +-------------------------------------+
  ```

Becavalier commented:
So, what the current status of this spec design? This feature had already implemented in Firefox Developer Edition.

codehag commented:
@Becavalier the spec hasn't been implemented in firefox yet, however we do have debugging capability for wasm (I believe it has been since late last year / early this year). This is distinct from the spec.

With regards to the spec, @dschuff will present a proposal for a subworking group to the web assembly working group and there will probably be a repository opened for iterating on the specification there.
